### PR TITLE
feat(batch): support hash based lookup join

### PR DESCRIPTION
**Pull request by chenzl25**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

- Original lookup join is inefficient when a large number of keys need to lookup, because its nested loop implementation. Hash based implementation can be 1000% faster than original one in some scenarios.
- This PR implement the hash based lookup join by reusing the hash join codes.
- It can speed up #5555 case from 10s + to 100ms+.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
#5555

### CODE REVIEW

File: src/batch/src/executor/join/lookup_join.rs [hunk 1]
```diff
@@ -317,263 +316,191 @@ impl<P: 'static + ProbeSideSourceBuilder> Executor for LookupJoinExecutor<P> {
...
-            for row_key in &row_keys {
+            for row_key in &groups {
                 self.probe_side_source.add_scan_range(row_key)?;
             }
+            let lookup_join_probe_child = self.probe_side_source.build_source().await?;
+
+            // NOTICE!!!
+            // Lookup join build side will become the probe side of hash join,
+            // while its probe side will become the build side of hash join.
+            let hash_join_probe_child = Box::new(BufferChunkExecutor::new(
```
Comment (BowenXiao1999) [inline src/batch/src/executor/join/lookup_join.rs:389]: This is quite confusing, maybe we should unfiy the naming 😢

**Comment by liurenjie1024** [comment]
This can improve performance when build side is quite large, but building hash table may affect latency for short queries. Did you test it in sysbench?

**Comment by liurenjie1024** [comment]
Another thing is the optimizer thinks lookup join keeps build side input's order, and this may change it. cc @st1page

**Comment by chenzl25** [comment]
> Another thing is the optimizer thinks lookup join keeps build side input's order, and this may change it. cc @st1page

Hash table is built by the lookup side of lookup join, so we can still preserve build side order.

**Comment by chenzl25** [comment]
> This can improve performance when build side is quite large, but building hash table may affect latency for short queries. Did you test it in sysbench?

Yes, it make no difference for sysbench single key index lookup latency. Lookup join latency is dominated by RPC.

File: src/batch/src/executor/join/lookup_join.rs [hunk 2]
```diff
@@ -317,263 +316,191 @@ impl<P: 'static + ProbeSideSourceBuilder> Executor for LookupJoinExecutor<P> {
...
+                hash_join_probe_data_types.clone(),
+                hash_join_build_data_types.clone(),
+            ]
+            .concat();
+
+            let mut build_side = Vec::new();
+            let mut build_row_count = 0;
+            #[for_await]
+            for build_chunk in hash_join_build_child.execute() {
+                // for build_chunk in chunk_list {
```
Comment (BowenXiao1999) [inline src/batch/src/executor/join/lookup_join.rs:409]: remove

**Comment by chenzl25** [comment]
BTW, I think 1000 records for an index lookup is quite normal, since we have no statistics about index columns and users  can build an index on whatever columns they like.

File: src/batch/src/executor/join/lookup_join.rs [hunk 3]
```diff
@@ -317,263 +316,191 @@ impl<P: 'static + ProbeSideSourceBuilder> Executor for LookupJoinExecutor<P> {
...
+                    JoinType::LeftSemi => {
+                        HashJoinExecutor::do_left_semi_join_with_non_equi_condition(params, cond)
+                    }
+                    JoinType::LeftAnti => {
+                        HashJoinExecutor::do_left_anti_join_with_non_equi_condition(params, cond)
+                    }
+                    JoinType::RightOuter
+                    | JoinType::RightSemi
+                    | JoinType::RightAnti
+                    | JoinType::FullOuter => unimplemented!(),
```
Comment (fuyufjh) [inline src/batch/src/executor/join/lookup_join.rs:470]: Have these been prohibited from frontend?

Will it be supported in the future? Or never going to support it?

**Review by liurenjie1024 — APPROVED** [review]

**Review by BowenXiao1999 — APPROVED** [review]

**Review by fuyufjh — APPROVED** [review]
LGTM for rest

File: src/batch/src/executor/join/lookup_join.rs (hunk 3 above)
Comment (chenzl25) [reply src/batch/src/executor/join/lookup_join.rs:470]: Yes, this kind of join type had already prohibited from frontend. For `Right` join, I think a proper way is to convert it to the `Left` join by optimizer. For `FullOuter` join type , lookup join can not provide this kind of semantics. So never going to support it.

File: src/batch/src/executor/join/lookup_join.rs (hunk 1 above)
Comment (chenzl25) [reply src/batch/src/executor/join/lookup_join.rs:398]: Leave it for later PR to do the rename stuff.

**Pull request merged**